packages/functions/src/db-connections/index.ts

```diff
@@ -0,0 +1,310 @@
+ * idle database connections to be removed from the pool.
+ *
+ * @param dbPool - The database pool object. The supported pool types are:
+ * - PostgreSQL (pg)
+ * - MySQL2
+ * - MariaDB
+ * - MongoDB
+ * - Redis (ioredis)
+ * - Cassandra (cassandra-driver)
+ * - OTHER: This method uses duck-typing to detect the pool type. Respectively you can
+ *   pass in any object with a compatible interface.
+ *
+ * @example
+ * ```ts
+ * const pgPool = new Pool({
+ *   connectionString: process.env.DATABASE_URL,
+ * });
+ * experimental_attachDatabasePool(pgPool);
+ * ```
+ *
+ * @experimental
+ */
+export function experimental_attachDatabasePool(dbPool: DbPool) {
+  if (idleTimeout) {
+    idleTimeoutResolve?.();
+    clearTimeout(idleTimeout);
+  }
+
+  // PostgreSQL, MySQL2, MariaDB - Listen for release events
+  if (
+    'on' in dbPool &&
+    dbPool.on &&
+    'options' in dbPool &&
+    'idleTimeoutMillis' in dbPool.options!
+  ) {
+    // This is PgPool
+    const pgPool = dbPool as PgPool;
+    pgPool.on!('release', () => {
+      if (DEBUG) {
+        console.log('Client released from pool');
+      }
+      waitUntilIdleTimeout(dbPool);
+    });
+    return;
+  } else if (
+    'on' in dbPool &&
+    dbPool.on &&
+    'config' in dbPool &&
+    dbPool.config &&
+    'connectionConfig' in dbPool.config
+  ) {
+    // This is MySQLPool
+    const mysqlPool = dbPool as MySQLPool;
+    mysqlPool.on!('release', () => {
+      if (DEBUG) {
+        console.log('MySQL client released from pool');
+      }
+      waitUntilIdleTimeout(dbPool);
+    });
+    return;
+  } else if (
+    'on' in dbPool &&
+    dbPool.on &&
+    'config' in dbPool &&
+    dbPool.config &&
+    'idleTimeout' in dbPool.config
+  ) {
+    // This is MySQL2Pool or MariaDBPool
+    const mysql2Pool = dbPool as MySQL2Pool | MariaDBPool;
+    mysql2Pool.on!('release', () => {
+      if (DEBUG) {
+        console.log('MySQL2/MariaDB client released from pool');
+      }
+      waitUntilIdleTimeout(dbPool);
+    });
+    return;
+  }
+
+  // MongoDB - Listen for connection pool events
+  if (
+    'on' in dbPool &&
+    dbPool.on &&
+    'options' in dbPool &&
+    dbPool.options &&
+    'maxIdleTimeMS' in dbPool.options
+  ) {
+    const mongoPool = dbPool as MongoDBPool;
+    mongoPool.on!('connectionCheckedOut', () => {
+      if (DEBUG) {
+        console.log('MongoDB connection checked out');
+      }
+      waitUntilIdleTimeout(dbPool);
+    });
+    return;
+  }
+
+  // Redis - Listen for connection events
+  if (
+    'on' in dbPool &&
+    dbPool.on &&
+    'options' in dbPool &&
+    dbPool.options &&
+    'socket' in dbPool.options
+  ) {
+    const redisPool = dbPool as RedisPool;
+    redisPool.on!('end', () => {
+      if (DEBUG) {
+        console.log('Redis connection ended');
+      }
+      waitUntilIdleTimeout(dbPool);
+    });
+    return;
+  }
+
+  throw new Error('Unsupported database pool type');
```

Comment:
Not sure about throwing an error, can we just log an error to not potentially completely destroy the instance?

Reply:
Hum, fair point. Let's roll